Objects/codeobject.c

```diff
@@ -1667,7 +1682,14 @@ code_new_impl(PyTypeObject *type, int argcount, int posonlyargcount,
 static void
 code_dealloc(PyCodeObject *co)
 {
+    assert(Py_REFCNT(co) == 0);
+    Py_SET_REFCNT(co, 1);
```

Comment:
Do you still want to move this to the event handler function? I imagine it would use a return value to indicate whether the object was resurrected.

Reply:
This adds complexity (particularly for dict watchers, where the notify func already has a return value, the new dict version tag.) It's doable, but I think it is only worth doing if it makes a detectable difference in pyperformance, and it doesn't seem to; this PR as-is is "1.00x faster" than the last merged commit from main branch.

Reply:
FWIW, another option is just 'if refcount is no longer 0, return'. I think it makes sense for code cleanliness more than performance reasons to put the refcount shenanigans inside the only branch that cares about them, but if you prefer it like this, that's fine.

Reply:
Hmm. The current version seems cleaner to me. The "branch that cares about" this is specifically the dealloc path, where refcount is 0. Ideally `*_dealloc` shouldn't be handing off a dead object to the black-box notify func (which is also used by many other paths that don't deal with dead objects), and the current version cleanly and clearly ensures that in the spot where we already know we have a dead object.

If it were a performance issue to do this refcount bump when there is no callback installed, then we could more tightly couple the notify func to the dealloc path, by making the notify func responsible for the fact that it might receive a dead object, when it's called with the dealloc event type. But that seems to me like a concession, reducing code cleanliness slightly in exchange for performance.